src/caffe/util/gpu_memory.cpp

```diff
@@ -0,0 +1,186 @@
+    break;
+  default:
+    break;
+  }
+  mode_ = NO_POOL;
+}
+
+void GPUMemoryManager::allocate(void** ptr, size_t size, cudaStream_t stream) {
+  CHECK((ptr) != NULL);
+  switch (mode_) {
+  case CUB_POOL:
+    if (cub_allocator->DeviceAllocate(ptr, size, stream) != cudaSuccess) {
+      int cur_device;
+      CUDA_CHECK(cudaGetDevice(&cur_device));
+      // free all cached memory (for all devices), synchrionize
+      cudaDeviceSynchronize();
+      cudaThreadSynchronize();
+      cub_allocator->FreeAllCached();
+      cudaDeviceSynchronize();
+      cudaThreadSynchronize();
+      // Refresh per-device saved values.
+      for (int i = 0; i < dev_info_.size(); ++i) {
+        // only query devices that were initialized
+        if (dev_info_[i].total_) {
+          update_dev_info(i);
+          // record which device caused cache flush
+          if (i == cur_device) {
+            dev_info_[i].flush_count_++;
+          }
+        }
+      }
+      // Retry once
+      CUDA_CHECK(cub_allocator->DeviceAllocate(ptr, size, stream));
+    }
+    // If retry succeeds we need to clean up last error
+    cudaGetLastError();
+    break;
+  default:
+    CUDA_CHECK(cudaMalloc(ptr, size));
+    break;
+  }
+}
+
+void GPUMemoryManager::deallocate(void* ptr, cudaStream_t stream) {
+  // allow for null pointer deallocation
+  if (!ptr) {
+    return;
+  }
+  switch (mode_) {
+  case CUB_POOL:
+    CUDA_CHECK(cub_allocator->DeviceFree(ptr));
+    break;
+  default:
+    CUDA_CHECK(cudaFree(ptr));
+    break;
+  }
+}
+
+void GPUMemoryManager::update_dev_info(int device) {
+  int initial_device;
+  CUDA_CHECK(cudaGetDevice(&initial_device));
+  if (device + 1 > dev_info_.size()) {
+    dev_info_.resize(device + 1);
+  }
+
+  CUDA_CHECK(cudaSetDevice(device));
+  cudaDeviceProp props;
+  CUDA_CHECK(cudaGetDeviceProperties(&props, device));
+  CUDA_CHECK(cudaMemGetInfo(&dev_info_[device].free_,
+      &dev_info_[device].total_));
+
+  VLOG_IF(1, debug_) << "cudaGetDeviceProperties: Mem = "
+                     << props.totalGlobalMem;
+  VLOG_IF(1, debug_) << "cudaMemGetInfo_[" << device << "]: Free="
+                     << dev_info_[device].free_ << " Total="
+                     << dev_info_[device].total_;
+
+  // Make sure we don't have more that total device memory
+  dev_info_[device].total_ = std::min(props.totalGlobalMem,
+      dev_info_[device].total_);
+  // Here we are adding existing 'busy' allocations to CUDA free memory
+  dev_info_[device].free_ = std::min(dev_info_[device].total_,
+      dev_info_[device].free_ + cub_allocator->cached_bytes[device].live);
+  CUDA_CHECK(cudaSetDevice(initial_device));
+}
+
+void GPUMemoryManager::InitMemory(const std::vector<int>& gpus, PoolMode m) {
+  mode_ = m;
+  switch (mode_) {
+  case CUB_POOL:
+    try {
+      // Just in case someone installed 'no cleanup' arena before
+      delete cub_allocator;
+      cub_allocator = new cub::CachingDeviceAllocator(2, 6, 22, (size_t) -1,
```

Comment:
Surely these parameters should not be hardcoded in the middle of this file, right?